doc/develop/languages/cpp/index.rst

```diff
@@ -121,5 +121,133 @@ library can select, from its config,
+configurations.
+Adding more constraints, the Zephyr project has demanding policies
+with respect to code style, compiler warnings, static analyzers and
+standard compliance (e.g.: MISRA).
+
+Put together, all these constraints can make writing header files very
+challenging. The purpose of this section is to document some best "header
+practices" and lessons learned in a Zephyr-specific context. While a lot
+of the information here is not Zephyr-specific, this section is not a
+substitute for knowledge of C/C++ standards, textbooks and other
+references.
+
+Testing
+-------
+
+Fortunately, the Zephyr project has an extensive test and CI
+infrastructure that provides coverage baselines, catches issues early,
+enforces policies and maintains such combinatorial explosions under some
+control. The ``tests/lib/cpp/cxx/`` are very useful in this context
+because their ``testcase.yaml`` configuration lets ``twister`` iterate
+quickly over a range of ``-std`` parameters: ``-std=c++98``,
+``-std=c++11``, etc.
+
+Keep in mind unused macros are not compiled.
+
+Designated initializers
+-----------------------
+
+Initialization macros are common in header files as they help reduce
+boilerplate code. C99 added initialization of ``struct`` and ``union``
+types by "designated" member names instead of a list of *bare*
+expressions. Some GCC versions support designated initializers even in
+their C90 mode.
+
+When used at a simple level, designated initializers are less
+error-prone, more readable and more flexible. On the other hand, C99
+allowed a surprisingly large and lax set of possibilities: designated
+initializers can be out of order, duplicated, "nested" (``.a.x =``),
+various other braces can be omitted, designated initializers and not can
+be mixed, etc.
+
+Twenty years later, C++20 added designated initializers to C++ but in
+much more restricted way; partly because a C++ ``struct`` is actually a
+``class``. As described in the C++ proposal number P0329 (which compares
+with C) or in any complete C++ reference, a mix is not allowed and
+initializers must be in order (gaps are allowed).
+
+Interestingly, the new restrictions in C++20 can cause ``gcc -std=c++20``
+to fail to compile code that successfully compiles with
+``gcc -std=c++17``. For example, ``gcc -std=c++17`` and older allow the
+C-style mix of initializers and bare expressions. This fails to compile
+with using ``gcc -std=c++20`` *with the same GCC version*.
+
+Recommendation: to maximize compatibility across different C and C++
+toolchains and standards, designated initializers in Zephyr header files
+should follow all C++20 rules and restrictions. Non-designated, pre-C99
+initialization offers more compatibility and is also allowed but
+designated initialization is the more readable and preferred code
+style. In any case, both styles must never be mixed in the same
+initializer.
+
+Warning: successful compilation is not the end of the incompatibility
```

Comment:
Should this be an actual `.. warning:` block, or is it just to make the text less dry?

Reply:
The latter. I don't think this is important enough to deserve the highlight of a warning block. Maybe make the "Warning" word bold?